Library/Homebrew/software_spec.rb

```diff
@@ -517,12 +522,23 @@ def compatible_locations?(tag: Utils::Bottles.tag)
 
     prefix = Pathname(cellar).parent.to_s
 
-    compatible_cellar = cellar == HOMEBREW_CELLAR.to_s
-    compatible_prefix = prefix == HOMEBREW_PREFIX.to_s
+    cellar_relocatable = cellar.size >= HOMEBREW_CELLAR.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
+    prefix_relocatable = prefix.size >= HOMEBREW_PREFIX.to_s.size && ENV["HOMEBREW_RELOCATE_BUILD_PREFIX"]
+
+    compatible_cellar = cellar == HOMEBREW_CELLAR.to_s || cellar_relocatable
+    compatible_prefix = prefix == HOMEBREW_PREFIX.to_s || prefix_relocatable
 
     compatible_cellar && compatible_prefix
   end
 
+  # Should the build prefix for the {Bottle} this {BottleSpecification} belongs to be relocated?
+  sig { params(tag: Utils::Bottles::Tag).returns(T::Boolean) }
+  def skip_prefix_relocation?(tag: Utils::Bottles.tag)
+    spec = collector.specification_for(tag)
+    relocatable_cellar = spec&.cellar == :any || spec&.cellar == :any_skip_relocation
+    relocatable_cellar || Homebrew.default_prefix?
```

Comment:
`spec.cellar == HOMEBREW_CELLAR.to_s` is probably more accurate rather than a default prefix check.

If `spec` is nil then you can check `tag.default_cellar`.

Honestly, the logic here sounds like it will probably be `compatible_locations?` but without the new size check so you can arrange the code as such.

Reply:
That's a good suggestion.  I'll rearrange this with `compatible_locations?` in mind.

Reply:
I've refactored `skip_prefix_relocation` to more closely follow `compatible_locaions?` now.  The code ends up looking kind of redundant so if anyone has suggestions on how to improve that I'd be happy to try.